File: Python/Nonlineq/FixPunktIterExer5P14.py

import numpy as np #Basic linear algebra
import matplotlib.pyplot as plt #Plotting


tolerance = 10**(-15)
def f(x):
    return x**2 -x -2

def f2(x):
    return np.sqrt(x +2)

def f3(x):
    return 1 + 2/x

def f4(x):
    return (x**2 + 2)/(2*x -1)

N=100 #start array of 100-ones 
y2=np.ones(N); y3=np.ones(N); y4=np.ones(N)
y2[0]=2.5; y3[0]=2.5; y4[0]=2.5 #start guess


iteration_fixpunkt = []
func_fixpunkt = np.ones(N)
k=0 #index for array + interation

#y2
while abs(y2[k]-y2[k-1])>tolerance: #Fixtpoint iteration
    y2[k+1] = f2(y2[k]) #
    func_fixpunkt[k] =f2(y2[k])
    k = k+1
    iteration_fixpunkt.append(k)
    if k > 100: break

func_fixpunkt=func_fixpunkt[:k] #remove extra ones from array after loop
plt.figure(0)
plt.plot(iteration_fixpunkt,func_fixpunkt, 'bo-',label="y2 konv.iter:"
         f'{k}' ",solution:" f'{func_fixpunkt[k-1]}')

#y3
iteration_fixpunkt = [] #reset
func_fixpunkt = np.ones(N)
k=0
while abs(y3[k]-y3[k-1])>tolerance: 
    y3[k+1] = f3(y3[k]) #
    func_fixpunkt[k] =f3(y3[k])
    k = k+1
    iteration_fixpunkt.append(k)
    if k > 100: break

func_fixpunkt=func_fixpunkt[:k] 
plt.plot(iteration_fixpunkt,func_fixpunkt, 'co--',label="y3 konv.iter:"
         f'{k}' ",solution:" f'{func_fixpunkt[k-1]}')

#y4
iteration_fixpunkt = [] #reset
func_fixpunkt = np.ones(N)
k=0
while abs(y4[k]-y4[k-1])>tolerance: 
    y4[k+1] = f4(y4[k]) #
    func_fixpunkt[k] =f4(y4[k])
    k = k+1
    iteration_fixpunkt.append(k)
    if k > 100: break

func_fixpunkt=func_fixpunkt[:k] 

plt.plot(iteration_fixpunkt,func_fixpunkt, 'ro-',label="y4 konv.iter:"
         f'{k}' ",solution:" f'{func_fixpunkt[k-1]}')
plt.xlabel('iteration')
plt.ylabel('function ')
plt.grid()
plt.legend()


n = -1.0
p = 3.0
x_exact = np.linspace(n,p,100);
func_exact=np.ones(len(x_exact))
M = len(x_exact)
for i in range(0,M):
    func_exact[i] =f(x_exact[i])
plt.figure(1)
plt.plot(x_exact,func_exact,"b-",label="x*np.sin(x)-1=0")
plt.xlabel('x')
plt.ylabel('function x**2-x-2=0')
plt.legend()

File: MATLAB/Nonlineq/FixPunktIterExer5P14.m
close all;

tolerance = 10e-15;

fun2 = @(x) sqrt(x+2);

fun3 = @(x)  1 + 2/x;
fun4 = @(x) (x^2 +2)/(2*x-1);


y2=0; y3=0;y4=0;
iteration_fixpunkt = 0;
func_fixpunkt = 0;
error_fixpunkt = 0;

%Initialiseringen: init guess
y2(1) = 2.5;
y3(1) = 2.5;
y4(1) = 2.5;
k=2;

iteration_fixpunkt(1) = 1;
func_fixpunkt(1) =fun2(y2(1));
y2(2) = fun2(y2(1));

% Main   fixpunkt's  iterations  
while abs( y2(k) - y2(k-1) ) > tolerance

y2(k+1) = fun2(y2(k));
iteration_fixpunkt(k) = k;
func_fixpunkt(k) =fun2(y2(k));

k = k+1;

if k == 100
break
end
  end

figure
 % plot(iteration_fixpunkt,error_fixpunkt, 'o b-','LineWidth',2)
 plot(iteration_fixpunkt,func_fixpunkt, 'o b-','LineWidth',2)
  
  xlabel('iteration')
  ylabel('function ')
grid on
 legend(['solution:',num2str(y2(k))]);
 %title(['Fixpunktsmetod, konvergens (iterations)'])

 
iterg2=k;
iteration_fixpunkt = 0;
func_fixpunkt = 0;
error_fixpunkt = 0;

k=2;

iteration_fixpunkt(1) = 1;
func_fixpunkt(1) =fun3(y3(1));
y3(2) = fun3(y3(1));


% Main   fixpunkt's  iterations  
while abs( y3(k) - y3(k-1) ) > tolerance

y3(k+1) = fun3(y3(k));
iteration_fixpunkt(k) = k;
func_fixpunkt(k) =fun3(y3(k));

k = k+1;

if k == 100
break
end
  end

hold on
%  figure
  plot(iteration_fixpunkt,func_fixpunkt, 'o c --','LineWidth',2)
  
  xlabel('iteration')
  ylabel('function ')
grid on
  legend(['solution:',num2str(y3(k))]);
% title(['Fixpunktsmetod, konvergens (iterations)'])

 iterg3=k;

iteration_fixpunkt = 0;
func_fixpunkt = 0;
error_fixpunkt = 0;


k=2;

iteration_fixpunkt(1) = 1;
func_fixpunkt(1) =fun4(y4(1));
y4(2) = fun4(y4(1));


% Main   fixpunkt's  iterations  
while abs( y4(k) - y4(k-1) ) > tolerance

y4(k+1) = fun4(y4(k));
iteration_fixpunkt(k) = k;
func_fixpunkt(k) =fun4(y4(k));

k = k+1;

if k == 100
break
end
  end

hold on
%  figure


  plot(iteration_fixpunkt,func_fixpunkt, 'o r  -','LineWidth',2)
  
  xlabel(['iterations: nr. iter.g2:',num2str(iterg2),', nr.iter. g3:', num2str(iterg3),' nr.iter. g4:', num2str(iterg4)])
  ylabel('function ')
grid on
  legend(['solution:',num2str(y4(k))]);
 title('Fixpunktsmetod, konvergens (iterations)')

 
iterg4=k;

 legend(['g_2 konv.iter.:',num2str(iterg2), ', solution:', num2str(y2(k))],['g_3 konv.iter.:',num2str(iterg3),', solution:', num2str(y3(k))],['g_4 konv.iter.:',num2str(iterg4), ', solution:', num2str(y4(k))]);
% legend(['g_2 konv.iter.:',num2str(y2(k))],['g_3 konv.iter.:',num2str(y3(k))],['g_4 konv.iter.:',num2str(y4(k))]);

 
%*********************************************************
n= -1.0;
p=3.0;
x_exact = linspace(n, p);
N = size(x_exact,2);

fun = @(x) x^2 -   x - 2;

 for i = 1:N
	    func_exact(i) =fun(x_exact(i));
 end
 
 
figure

plot(x_exact,func_exact, ' b-','LineWidth',2)
 legend(' x^2 - x - 2= 0 ');

  xlabel('  x')
  ylabel('function x^2 - x -2 = 0  ')

grid on


